modules/javafx.graphics/src/main/java/javafx/scene/text/LayoutInfo.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package javafx.scene.text;
+
+import java.util.List;
+import javafx.geometry.Rectangle2D;
+
+/**
+ * Provides a view into the text layout used in a {@code Text} or a {@code TextFlow} node,
+ * with purpose of querying the details of the layout such as break up of the text into lines,
+ * as well as geometry of other shapes derived from the layout (selection, underline, etc.).
+ * <p>
+ * The information obtained via this object may change after the next layout cycle, which may come as a result
+ * of actions such as resizing of the container, or modification of certain properties.
+ * For example updating the text or the font might change the layout, but a change of color would not.
+ *
+ * @since 24
+ */
+public sealed abstract class LayoutInfo permits com.sun.javafx.text.PrismLayoutInfo {
+    /**
+     * Constructor for subclasses to call.
+     */
+    protected LayoutInfo() {
+    }
+
+    /**
+     * Returns the logical bounds of the layout:
+     * <ul>
+     * <li>{@code minX} is always zero
+     * <li>{@code minY} is the ascent of the first line (negative)
+     * <li>{@code width} the width of the widest line
+     * <li>{@code height} the sum of all lines height
+     * </ul>
+     *
+     * @return the layout bounds
+     */
+    public abstract Rectangle2D getBounds();
+
+    /**
+     * Returns the number of text lines in the layout.
+     * @return the number of text lines
+     */
+    public abstract int getTextLineCount();
+
+    /**
+     * Returns the list of text lines in the layout.
+     *
+     * @return the list of {@code TextLineInfo} objects
+     */
+    public abstract List<TextLineInfo> getTextLines();
```

Comment:
Is the list immutable? If so (which I think it should be), can you specify that?

Reply:
A new instance gets created for each caller, we don't care what they do to it.  Maybe we should say that?

From the API purity perspective, it could be, but an immutable list would incur a copy overhead, do we really need that?

It also looks like there is no easy way of creating an immutable wrapper similar to JDK's List12, so we'd need to invent one.

(Note: the original design called for an array, but List was recommended because of stream/etc)

Reply:
You can use `Collections.unmodifiableList(List)` to wrap the list without copying.